core/src/main/java/hudson/model/Queue.java

```diff
@@ -377,6 +396,9 @@ public synchronized void load() {
                             if(item.task==null)
                                 continue;   // botched persistence. throw this one away
 
+                            // backward compatibility - in case the serialized queue state was
+                            // a List of items (as of old), then the maxId is the max of the items in
+                            // the persisted queue.
```

Comment:
So there was always a loop in there performing checks on the persisted list items ala

```
for (Object o : state.items) {
    if (o instanceof Task) {
        // backward compatibility
        schedule((Task)o, 0);
    } else if (o instanceof Item) {
        Item item = (Item)o;
        if(item.task==null)
            continue;   // botched persistence. throw this one away

        maxId = Math.max(maxId, item.id);
        if (item instanceof WaitingItem) {
            item.enter(this);
        } else if (item instanceof BlockedItem) {
            item.enter(this);
        } else if (item instanceof BuildableItem) {
            item.enter(this);
        } else {
            throw new IllegalStateException("Unknown item type! " + item);
        }
    } // this conveniently ignores null
}
```

i.e. there was more than the maxId tango going on. I intentionally did not touch that, other than to add the comment before `maxId = Math.max(maxId, item.id);`.

Sure the `maxId` stuff can be moved out of the loop, but I'd be for leaving the other code as is and leaving those checks as they were, whether we have a `State` or a `List`.